### DISCUSSION TITLE
Adding Elements to DOCTYPE

### DISCUSSION THREAD

**Discussion by ffreality** [zeux/pugixml#579] [Q&A]
Is there any sample about how can I create elementy types for DOCTYPE like this in runtime ?
https://www.javaskool.com/xml-document-type-definition/

I tried these
https://github.com/FF-Plugins-Docs/FF_PugiXml/blob/main/Source/FF_PugiXml/Private/FF_PugiXmlBPLibrary.cpp#L137
https://github.com/FF-Plugins-Docs/FF_PugiXml/blob/main/Source/FF_PugiXml/Private/FF_PugiXmlBPLibrary.cpp#L173

I tried second code after adding all of nodes (before saving it) if in case it automatically creates elements.
Also I tried to give attributes and childrens to DOCTYPE but it didn't work.

Right now my XML is like this.

```
<?xml version="1.0" encoding="utf-8" standalone="no"?>
<!DOCTYPE UnrealAttributes>
<UE5_PUGIXML xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <COMPANIES>
    <Company_VR Language="C++" Tool="UE5">Frozen Forest</Company_VR>
    <Company_Edu Language="Kotlin" Tool="IoT">Talkido</Company_Edu>
    <Company_Web Language="Python" Tool="AWS">Monolib</Company_Web>
  </COMPANIES>
</UE5_PUGIXML>
```

**Comment by zeux** [comment]
The code looks correct on first glance. Your example XML does have a `!DOCTYPE` node in the output, are you expecting to see something else?

One note is that pugixml only supports a simplified way to model DOCTYPE that requires custom assembly of the contents of the DOCTYPE including any nested groups. For example, to reproduce the following more complex definition:

```
<!DOCTYPE dtd-name
[
<!ELEMENT element-name (element-content type)>
<!ATTLIST element-name attribute-name attribute-type default-value>
]>
```

you'd have to set the doctype value to the full string with raw contents, something like

```
std::string dtd = "dtd-name\n";
dtd += "[\n";
dtd += "<!ELEMENT element-name (element-content type)>\n";
dtd += "<!ATTLIST element-name attribute-name attribute-type default-value>\n";
dtd += "]";

dtd_node.set_value(dtd.c_str());
```

**Comment by ffreality** [comment]
Oh okay. So, I need to append strings to get result which I want and use that final form as value.
Thanks !
